File: test/musicGame/geometry/RectangleTest.java
package musicGame.geometry;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

import org.junit.Test;

public class RectangleTest {

  private void testRectIntersectionWithParameters(
      int x1, int y1, int width1, int height1,
      int x2, int y2, int width2, int height2,
      float ix, float iy, float iwidth, float iheight) {
    Rectangle rect1 = new Rectangle(x1, y1, width1, height1);
    Rectangle rect2 = new Rectangle(x2, y2, width2, height2);

    Rectangle inter = rect1.intersection(rect2);
    assertThat(inter.getX(), is(ix));
    assertThat(inter.getY(), is(iy));
    assertThat(inter.getWidth(), is(iwidth));
    assertThat(inter.getHeight(), is(iheight));
  }

  @Test
  public void getIntersectionShouldReturnIntersectionOfTwoRectangles() {
    testRectIntersectionWithParameters(10, 10, 50, 30, 0, 10, 20, 10, 10, 10, 10, 10);
    testRectIntersectionWithParameters(10, 20, 50, 30, 5, 26, 100, 30, 10, 26, 50, 24);
  }

  @Test
  public void getIntersectionShouldReturnSmallestRectangleIfCompletelyInLargerOne() {
    testRectIntersectionWithParameters(0, 0, 100, 100, 0, 0, 50, 50, 0, 0, 50, 50);
    testRectIntersectionWithParameters(0, 0, 100, 100, 0, 0, 100, 100, 0, 0, 100, 100);
    testRectIntersectionWithParameters(265, 15, 490, 100, 265, 15, 490, 100, 265, 15, 490, 100);
    testRectIntersectionWithParameters(0, 0, 100, 100, 10, 10, 50, 50, 10, 10, 50, 50);
    testRectIntersectionWithParameters(10, 10, 50, 50, 0, 0, 100, 100, 10, 10, 50, 50);
  }

  @Test
  public void getIntersectionShouldReturnZeroWidthHeightRectangleIfNoIntersection() {
    testRectIntersectionWithParameters(
        0, 0, 10, 15,
        100, 1000, 10, 15,
        0, 0, 0, 0);
    testRectIntersectionWithParameters(
        0, 0, 10, 10,
        10, 0, 10, 10,
        0, 0, 0, 0);
    testRectIntersectionWithParameters(
        0, 0, 10, 10,
        0, 10, 10, 10,
        0, 0, 0, 0);
  }

  @Test
  public void horizontalCollisionDirectionShouldReturnTrueIfRightOfCenter() {
    Rectangle rect1 = new Rectangle(0, 0, 20, 20);
    Rectangle rect2 = new Rectangle(10, 0, 20, 20);
    assertThat(rect1.horizontalCollisionDirection(rect2), is(true));
    rect2 = new Rectangle(1, 20, 20, 20);
    assertThat(rect1.horizontalCollisionDirection(rect2), is(true));
  }

  @Test
  public void horizontalCollisionDirectionShouldReturnFalseIfLeftOfCenter() {
    Rectangle rect1 = new Rectangle(0, 0, 20, 20);
    Rectangle rect2 = new Rectangle(-1, 0, 20, 20);
    assertThat(rect1.horizontalCollisionDirection(rect2), is(false));
    rect2 = new Rectangle(0, 20, 20, 20);
    assertThat(rect1.horizontalCollisionDirection(rect2), is(false));
  }

  @Test
  public void verticalCollisionDirectionShouldReturnTrueIfBelowCenter() {
    Rectangle rect1 = new Rectangle(0, 0, 20, 20);
    Rectangle rect2 = new Rectangle(-1, 550, 20, 20);
    assertThat(rect1.verticalCollisionDirection(rect2), is(true));
    rect2 = new Rectangle(0, 1, 20, 20);
    assertThat(rect1.verticalCollisionDirection(rect2), is(true));
  }

  @Test
  public void verticalCollisionDirectionShouldReturnFalseIfAboveCenter() {
    Rectangle rect1 = new Rectangle(0, 0, 20, 20);
    Rectangle rect2 = new Rectangle(-1, -42, 20, 20);
    assertThat(rect1.verticalCollisionDirection(rect2), is(false));
    rect2 = new Rectangle(0, 0, 20, 20);
    assertThat(rect1.verticalCollisionDirection(rect2), is(false));
  }

  @Test
  public void intersectsShouldReturnFalseWhenTouchingSameLine() {
    Rectangle rect1 = new Rectangle(0, 0, 20, 20);
    Rectangle rect2 = new Rectangle(20, 0, 20, 20);
    Rectangle rect3 = new Rectangle(0, 20, 20, 20);

    assertThat(rect1.intersects(rect2), is(false));
    assertThat(rect1.intersects(rect3), is(false));
  }
}

File: javatests/believe/geometry/RectangleTest.kt
package believe.geometry

import com.google.common.truth.Truth.assertThat
import org.junit.jupiter.api.Test

internal class RectangleTest {
    @Test
    fun rectangle_fromProperties_buildsCorrectRectangle() {
        with(rectangle(1.2f, 3.4f, 5.6f, 7.8f)) {
            assertThat(x).isEqualTo(1.2f)
            assertThat(y).isEqualTo(3.4f)
            assertThat(width).isEqualTo(5.6f)
            assertThat(height).isEqualTo(7.8f)
        }
    }

    @Test
    fun rectangle_fromSlickRectangle_buildsCorrectRectangle() {
        with(rectangle(org.newdawn.slick.geom.Rectangle(1.2f, 3.4f, 5.6f, 7.8f))) {
            assertThat(x).isEqualTo(1.2f)
            assertThat(y).isEqualTo(3.4f)
            assertThat(width).isEqualTo(5.6f)
            assertThat(height).isEqualTo(7.8f)
        }
    }
    @Test
    fun mutableRectangle_fromProperties_buildsCorrectRectangle() {
        with(mutableRectangle(1.2f, 3.4f, 5.6f, 7.8f)) {
            assertThat(x).isEqualTo(1.2f)
            assertThat(y).isEqualTo(3.4f)
            assertThat(width).isEqualTo(5.6f)
            assertThat(height).isEqualTo(7.8f)
        }
    }

    @Test
    fun mutableRectangle_fromSlickRectangle_buildsCorrectRectangle() {
        with(mutableRectangle(org.newdawn.slick.geom.Rectangle(1.2f, 3.4f, 5.6f, 7.8f))) {
            assertThat(x).isEqualTo(1.2f)
            assertThat(y).isEqualTo(3.4f)
            assertThat(width).isEqualTo(5.6f)
            assertThat(height).isEqualTo(7.8f)
        }
    }

    @Test
    fun asSlickRectangle_returnsCorrectSlickRectangle() {
        with(rectangle(1f, 2f, 3f, 4f).asSlickRectangle()) {
            assertThat(x).isEqualTo(1f)
            assertThat(y).isEqualTo(2f)
            assertThat(width).isEqualTo(3f)
            assertThat(height).isEqualTo(4f)
        }
    }

    @Test
    fun intersectionShouldReturnIntersectionOfTwoRectangles() {
        with(rectangle(10f, 10f, 50f, 30f).intersection(rectangle(0f, 10f, 20f, 10f))) {
            assertThat(x).isEqualTo(10f)
            assertThat(y).isEqualTo(10f)
            assertThat(width).isEqualTo(10f)
            assertThat(height).isEqualTo(10f)
        }
        with(rectangle(10f, 20f, 50f, 30f).intersection(rectangle(5f, 26f, 100f, 30f))) {
            assertThat(x).isEqualTo(10f)
            assertThat(y).isEqualTo(26f)
            assertThat(width).isEqualTo(50f)
            assertThat(height).isEqualTo(24f)
        }
    }

    @Test
    fun intersectionShouldReturnSmallestRectangleIfCompletelyInLargerOne() {
        with(rectangle(0f, 0f, 100f, 100f).intersection(rectangle(0f, 0f, 50f, 50f))) {
            assertThat(x).isEqualTo(0f)
            assertThat(y).isEqualTo(0f)
            assertThat(width).isEqualTo(50f)
            assertThat(height).isEqualTo(50f)
        }
        with(rectangle(0f, 0f, 100f, 100f).intersection(rectangle(0f, 0f, 100f, 100f))) {
            assertThat(x).isEqualTo(0f)
            assertThat(y).isEqualTo(0f)
            assertThat(width).isEqualTo(100f)
            assertThat(height).isEqualTo(100f)
        }
        with(rectangle(265f, 15f, 490f, 100f).intersection(rectangle(265f, 15f, 490f, 100f))) {
            assertThat(x).isEqualTo(265f)
            assertThat(y).isEqualTo(15f)
            assertThat(width).isEqualTo(490f)
            assertThat(height).isEqualTo(100f)
        }
        with(rectangle(0f, 0f, 100f, 100f).intersection(rectangle(10f, 10f, 50f, 50f))) {
            assertThat(x).isEqualTo(10f)
            assertThat(y).isEqualTo(10f)
            assertThat(width).isEqualTo(50f)
            assertThat(height).isEqualTo(50f)
        }
        with(rectangle(10f, 10f, 50f, 50f).intersection(rectangle(0f, 0f, 100f, 100f))) {
            assertThat(x).isEqualTo(10f)
            assertThat(y).isEqualTo(10f)
            assertThat(width).isEqualTo(50f)
            assertThat(height).isEqualTo(50f)
        }
    }

    @Test
    fun intersectionShouldReturnZeroWidthHeightRectangleIfNoIntersection() {
        with(rectangle(0f, 0f, 10f, 15f).intersection(rectangle(100f, 1000f, 10f, 15f))) {
            assertThat(x).isEqualTo(0f)
            assertThat(y).isEqualTo(0f)
            assertThat(width).isEqualTo(0f)
            assertThat(height).isEqualTo(0f)
        }
        with(rectangle(0f, 0f, 10f, 10f).intersection(rectangle(10f, 0f, 10f, 10f))) {
            assertThat(x).isEqualTo(0f)
            assertThat(y).isEqualTo(0f)
            assertThat(width).isEqualTo(0f)
            assertThat(height).isEqualTo(0f)
        }
        with(rectangle(0f, 0f, 10f, 10f).intersection(rectangle(0f, 10f, 10f, 10f))) {
            assertThat(x).isEqualTo(0f)
            assertThat(y).isEqualTo(0f)
            assertThat(width).isEqualTo(0f)
            assertThat(height).isEqualTo(0f)
        }
    }

    @Test
    fun horizontalCollisionDirectionShouldReturnTrueIfRightOfCenter() {
        val rect1 = rectangle(0f, 0f, 20f, 20f)
        var rect2 = rectangle(10f, 0f, 20f, 20f)

        assertThat(rect1.horizontalCollisionDirection(rect2)).isTrue()
        rect2 = rectangle(1f, 20f, 20f, 20f)
        assertThat(rect1.horizontalCollisionDirection(rect2)).isTrue()
    }

    @Test
    fun horizontalCollisionDirectionShouldReturnFalseIfLeftOfCenter() {
        val rect1 = rectangle(0f, 0f, 20f, 20f)
        var rect2 = rectangle(-1f, 0f, 20f, 20f)

        assertThat(rect1.horizontalCollisionDirection(rect2)).isFalse()
        rect2 = rectangle(0f, 20f, 20f, 20f)
        assertThat(rect1.horizontalCollisionDirection(rect2)).isFalse()
    }

    @Test
    fun verticalCollisionDirectionShouldReturnTrueIfBelowCenter() {
        val rect1 = rectangle(0f, 0f, 20f, 20f)
        var rect2 = rectangle(-1f, 550f, 20f, 20f)

        assertThat(rect1.verticalCollisionDirection(rect2)).isTrue()
        rect2 = rectangle(0f, 1f, 20f, 20f)
        assertThat(rect1.verticalCollisionDirection(rect2)).isTrue()
    }

    @Test
    fun verticalCollisionDirectionShouldReturnFalseIfAboveCenter() {
        val rect1 = rectangle(0f, 0f, 20f, 20f)
        var rect2 = rectangle(-1f, -42f, 20f, 20f)

        assertThat(rect1.verticalCollisionDirection(rect2)).isFalse()
        rect2 = rectangle(0f, 0f, 20f, 20f)
        assertThat(rect1.verticalCollisionDirection(rect2)).isFalse()
    }

    @Test
    fun intersectsShouldReturnFalseWhenTouchingSameLine() {
        val rect1 = rectangle(0f, 0f, 20f, 20f)
        val rect2 = rectangle(20f, 0f, 20f, 20f)
        val rect3 = rectangle(0f, 20f, 20f, 20f)

        assertThat(rect1.intersects(rect2)).isFalse()
        assertThat(rect1.intersects(rect3)).isFalse()
    }

    @Test
    fun minX_returnsMinimumX() {
        assertThat(rectangle(10f, 20f, 100f, 1000f).minX).isEqualTo(10f)
    }

    @Test
    fun maxX_returnsMaximumX() {
        assertThat(rectangle(10f, 20f, 100f, 1000f).maxX).isEqualTo(110f)
    }

    @Test
    fun minY_returnsMinimumY() {
        assertThat(rectangle(10f, 20f, 100f, 1000f).minY).isEqualTo(20f)
    }

    @Test
    fun maxY_returnsMaximumY() {
        assertThat(rectangle(10f, 20f, 100f, 1000f).maxY).isEqualTo(1020f)
    }

    @Test
    fun centerX_returnsCenterX() {
        assertThat(rectangle(10f, 20f, 100f, 1000f).centerX).isEqualTo(60f)
    }

    @Test
    fun centerY_returnsCenterY() {
        assertThat(rectangle(10f, 20f, 100f, 1000f).centerY).isEqualTo(520f)
    }

    @Test
    fun center_returnsCenterCoordinates() {
        val rectangle = rectangle(10f, 10f, 100f, 100f)

        with(rectangle.center) {
            assertThat(x).isEqualTo(rectangle.centerX)
            assertThat(y).isEqualTo(rectangle.centerY)
        }
    }

    @Test
    fun center_rectangleIsMutated_returnsCenterCoordinates() {
        val rectangle = mutableRectangle(10f, 10f, 100f, 100f)
        rectangle.x = 100f

        with(rectangle.center) {
            assertThat(x).isEqualTo(rectangle.centerX)
            assertThat(y).isEqualTo(rectangle.centerY)
        }
    }
}